src/App.tsx

```diff
@@ -1,26 +1,254 @@
-  if (!USER_ID) {
-    return <UserWarning />;
-  }
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [filter, setFilter] = useState<FilterOptions>(FilterOptions.ALL);
+  const [errorMessage, setErrorMessage] = useState(ErrorMessages.NO_ERRORS);
+  const [isSubmitting, setIsSubmitting] = useState(false);
+  const [inputValue, setInputValue] = useState('');
+  const [tempTodo, setTempTodo] = useState<Todo | null>(null);
+  const [shouldFocus, setShouldFocus] = useState(true);
+  const [loadingTodosCount, setLoadingTodosCount] = useState<number[]>([]);
+
+  const inputRef = useRef<HTMLInputElement | null>(null);
+
+  useEffect(() => {
+    if (inputRef.current) {
+      inputRef.current.focus();
+    }
+  }, [todos]);
+
+  const handleInputChange = (event: React.ChangeEvent<HTMLInputElement>) => {
+    setInputValue(event.target.value);
+  };
+
+  const handleSubmit = (event: FormEvent<HTMLFormElement>) => {
+    event.preventDefault();
+
+    if (inputValue.trim() === '') {
+      setErrorMessage(ErrorMessages.EMPTY_TITLE);
+
+      return;
+    }
+
+    const temporaryTodo: Todo = {
+      id: 0,
+      title: inputValue.trim(),
+      userId: USER_ID,
+      completed: false,
+    };
+
+    setTempTodo(temporaryTodo);
+    setIsSubmitting(true);
+    setShouldFocus(false);
+    setLoadingTodosCount(currentCount => [...currentCount, 0]);
+
+    addTodo({ title: inputValue.trim(), userId: USER_ID, completed: false })
+      .then((createdTodo: Todo) => {
+        setTodos((currentTodos: Todo[]) => [...currentTodos, createdTodo]);
+        setInputValue('');
+        setErrorMessage(ErrorMessages.NO_ERRORS);
+        setShouldFocus(true);
+
+        if (inputRef.current) {
+          inputRef.current.focus();
+        }
+      })
+      .catch(() => {
+        setShouldFocus(true);
+        setErrorMessage(ErrorMessages.ADDING_ERROR);
+
+        if (inputRef.current) {
+          inputRef.current.focus();
+        }
+      })
+      .finally(() => {
+        if (inputRef.current) {
+          inputRef.current.disabled = false;
+          inputRef.current.focus();
+        }
+
+        setIsSubmitting(false);
+        setTempTodo(null);
+        setLoadingTodosCount(currentCount =>
+          currentCount.filter(todoId => todoId !== 0),
+        );
+      });
+  };
+
+  const handleTodoDelete = useCallback(
+    (todoId: number) => {
+      setIsSubmitting(true);
+      setLoadingTodosCount(current => [...current, todoId]);
+      deleteTodos(todoId)
+        .then(() => {
+          setTodos(currentTodos =>
+            currentTodos.filter(todo => todo.id !== todoId),
+          );
+        })
+        .catch(() => {
+          setErrorMessage(ErrorMessages.DELETING_ERROR);
+        })
+        .finally(() => {
+          setIsSubmitting(false);
+          setLoadingTodosCount(current =>
+            current.filter(deletingTodoId => todoId !== deletingTodoId),
+          );
+        });
+    },
+    [todos],
```

Comment:
Remove `todos `from the dependency array in the `useCallback `hooks where it's not directly used.